runtime/src/iree/modules/vmvx/module.c

```diff
@@ -61,13 +61,349 @@ iree_vmvx_module_free_state(void* self, iree_vm_module_state_t* module_state) {
 }
 
 //===----------------------------------------------------------------------===//
-// TODO
+// Private shims not needed by anything else.
 //===----------------------------------------------------------------------===//
 
-// Placeholder to make the function pointer arrays happy (they can't be empty).
-IREE_VM_ABI_EXPORT(iree_vmvx_module_placeholder,  //
-                   iree_vmvx_module_state_t,      //
-                   v, v) {
+IREE_VM_ABI_FIXED_STRUCT(irIIII, {
+  int32_t i0;
+  iree_vm_ref_t r1;
+  int64_t i2;
+  int64_t i3;
+  int64_t i4;
+  int64_t i5;
+});
+
+IREE_VM_ABI_FIXED_STRUCT(rIIrIIrIIIIIffi, {
+  iree_vm_ref_t r0;
+  int64_t i1;
+  int64_t i2;
+  iree_vm_ref_t r3;
+  int64_t i4;
+  int64_t i5;
+  iree_vm_ref_t r6;
+  int64_t i7;
+  int64_t i8;
+  int64_t i9;
+  int64_t i10;
+  int64_t i11;
+  float f12;
+  float f13;
+  int32_t i14;
+});
+
+// rIIIrIIIII
+IREE_VM_ABI_FIXED_STRUCT(rIIIrIIIII, {
+  iree_vm_ref_t r0;
+  int64_t i1;
+  int64_t i2;
+  int64_t i3;
+  iree_vm_ref_t r4;
+  int64_t i5;
+  int64_t i6;
+  int64_t i7;
+  int64_t i8;
+  int64_t i9;
+});
+
+// Since these are private/unlikely to be general, mark as static. If there
+// ever becomes one of these in the common set, the compiler will complain.
+static IREE_VM_ABI_DEFINE_SHIM(irIIII, v);
+static IREE_VM_ABI_DEFINE_SHIM(rIIIrIIIII, v);
+static IREE_VM_ABI_DEFINE_SHIM(rIIIrIIIrIIIII, v);
+static IREE_VM_ABI_DEFINE_SHIM(rIIrIIrIIIIIffi, v);
+
+static iree_status_t buffer_map(const iree_vm_buffer_t* buffer,
```

Comment:
They are pretty different... I was aiming these at reducing LOC for these kinds of cases. Can take a closer look, but I also expect to pretty wholesale rip/replace/macro-ize/etc this code in a future iteration. How it is done does not scale (to either microkernel count or the integrations needed).

Reply:
They look like verbatim copies to me :)
The only thing the shims below do is deref as a convenience but having that + a call to the base impl as a static inline function in here seems like it'd be fine - just wrap your junk around the common check.

It's not the biggest deal (there's worse duplication in other places today) but you've got me curious now because I don't understand the reticence: what do you think you could remove from these? If there's checks the buffer.c implementations are missing it's probably something we should have in there _or_ there's something shady planned that will make it really easy to cause SIGBUS or out of bounds issues. I don't buy the scaling issue - if the cost of 2 per operand bit twiddles, an addition, and a 100% predictable branch are significant to a microkernel then the granularity of the microkernel is wrong - every scalar load/store today is going through this path and it will be orders of magnitude better with vmvx :)

Reply:
> They look like verbatim copies to me :) The only thing the shims below do is deref as a convenience but having that + a call to the base impl as a static inline function in here seems like it'd be fine - just wrap your junk around the common check.
> 
> It's not the biggest deal (there's worse duplication in other places today) but you've got me curious now because I don't understand the reticence: what do you think you could remove from these? If there's checks the buffer.c implementations are missing it's probably something we should have in there _or_ there's something shady planned that will make it really easy to cause SIGBUS or out of bounds issues. I don't buy the scaling issue - if the cost of 2 per operand bit twiddles, an addition, and a 100% predictable branch are significant to a microkernel then the granularity of the microkernel is wrong - every scalar load/store today is going through this path and it will be orders of magnitude better with vmvx :)

They are a bit different in signatures. Let me take a pass through and try to clean this all up tomorrow. Just very short on time and was trying to phase some typing... But I have all day tomorrow free in theory.